File: glean-core/csharp/GleanTests/Metrics/StringMetricTypeTest.cs
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.

using Mozilla.Glean.Testing;
using System;
using System.IO;
using Xunit;
using static Mozilla.Glean.Glean;

namespace Mozilla.Glean.Tests.Metrics
{
    public class StringMetricTypeTest
    {
        public StringMetricTypeTest()
        {
            // Get a random test directory just for this single test.
            string tempDataDir = Path.Combine(Path.GetTempPath(), Path.GetRandomFileName());

            // In xUnit, the constructor will be called before each test. This
            // feels like a natural place to initialize / reset Glean.
            GleanInstance.Reset(
                applicationId: "org.mozilla.csharp.tests",
                applicationVersion: "1.0-test",
                uploadEnabled: true,
                configuration: new Configuration(),
                dataDir: tempDataDir
                );
        }

        [Fact]
        public void APISavesToStorage()
        {
            Private.StringMetricType stringMetric = new Private.StringMetricType(
                category: "telemetry",
                disabled: false,
                lifetime: Private.Lifetime.Application,
                name: "string_metric",
                sendInPings: new string[] { "store1" }
            );

            // Record two strings of the same type, with a little delay.
            stringMetric.Set("value");

            // Check that data was properly recorded.
            Assert.True(stringMetric.TestHasValue());
            Assert.Equal("value", stringMetric.TestGetValue());

            stringMetric.Set("overriddenValue");

            // Check that data was properly recorded.
            Assert.True(stringMetric.TestHasValue());
            Assert.Equal("overriddenValue", stringMetric.TestGetValue());
        }

        [Fact]
        public void DisabledstringsMustNotRecordData()
        {
            Private.StringMetricType stringMetric = new Private.StringMetricType(
                category: "telemetry",
                disabled: true,
                lifetime: Private.Lifetime.Application,
                name: "string_metric",
                sendInPings: new string[] { "store1" }
            );

            // Attempt to store the string.
            stringMetric.Set("value");
            // Check that nothing was recorded.
            Assert.False(stringMetric.TestHasValue(), "Strings must not be recorded if they are disabled");
        }

        [Fact]
        public void TestGetValueThrows()
        {
            Private.StringMetricType stringMetric = new Private.StringMetricType(
                category: "telemetry",
                disabled: true,
                lifetime: Private.Lifetime.Application,
                name: "string_metric",
                sendInPings: new string[] { "store1" }
            );
            Assert.Throws<NullReferenceException>(() => stringMetric.TestGetValue());
        }

        [Fact]
        public void APISavesToSecondaryPings()
        {
            Private.StringMetricType stringMetric = new Private.StringMetricType(
                category: "telemetry",
                disabled: false,
                lifetime: Private.Lifetime.Application,
                name: "string_metric",
                sendInPings: new string[] { "store1", "store2" }
            );

            // Record two strings of the same type, with a little delay.
            stringMetric.Set("value");

            // Check that data was properly recorded in the second ping.
            Assert.True(stringMetric.TestHasValue("store2"));
            Assert.Equal("value", stringMetric.TestGetValue("store2"));

            stringMetric.Set("overriddenValue");
            // Check that data was properly recorded in the second ping.
            Assert.True(stringMetric.TestHasValue("store2"));
            Assert.Equal("overriddenValue", stringMetric.TestGetValue("store2"));
        }

        [Fact]
        public void SettingALongStringRecordsAnError()
        {
            Private.StringMetricType stringMetric = new Private.StringMetricType(
                category: "telemetry",
                disabled: false,
                lifetime: Private.Lifetime.Application,
                name: "string_metric",
                sendInPings: new string[] { "store1", "store2" }
            );

            stringMetric.Set(new string('3', 110));

            Assert.Equal(1, stringMetric.TestGetNumRecordedErrors(ErrorType.InvalidOverflow));
        }
    }
}

File: glean-core/android/src/test/java/mozilla/telemetry/glean/private/StringMetricTypeTest.kt
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
* file, You can obtain one at http://mozilla.org/MPL/2.0/. */

/* This file is based on the tests in the Glean android-components implentation.
 *
 * Care should be taken to not reorder elements in this file so it will be easier
 * to track changes in Glean android-components.
 */

package mozilla.telemetry.glean.private

import androidx.test.core.app.ApplicationProvider
import androidx.test.ext.junit.runners.AndroidJUnit4
import mozilla.telemetry.glean.testing.ErrorType
import mozilla.telemetry.glean.testing.GleanTestRule
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class StringMetricTypeTest {

    @get:Rule
    val gleanRule = GleanTestRule(ApplicationProvider.getApplicationContext())

    @Test
    fun `The API saves to its storage engine`() {
        // Define a 'stringMetric' string metric, which will be stored in "store1"
        val stringMetric = StringMetricType(
            CommonMetricData(
                disabled = false,
                category = "telemetry",
                lifetime = Lifetime.APPLICATION,
                name = "string_metric",
                sendInPings = listOf("store1")
            )
        )

        // Record two strings of the same type, with a little delay.
        stringMetric.set("value")

        // Check that data was properly recorded.
        assertEquals("value", stringMetric.testGetValue())

        stringMetric.set("overriddenValue")
        // Check that data was properly recorded.
        assertEquals("overriddenValue", stringMetric.testGetValue())
    }

    @Test
    fun `disabled strings must not record data`() {
        // Define a 'stringMetric' string metric, which will be stored in "store1". It's disabled
        // so it should not record anything.
        val stringMetric = StringMetricType(
            CommonMetricData(
                disabled = true,
                category = "telemetry",
                lifetime = Lifetime.APPLICATION,
                name = "stringMetric",
                sendInPings = listOf("store1")
            )
        )

        // Attempt to store the string.
        stringMetric.set("value")
        // Check that nothing was recorded.
        assertNull(
            "Strings must not be recorded if they are disabled",
            stringMetric.testGetValue()
        )
    }

    @Test
    fun `testGetValue() returns null if nothing is stored`() {
        val stringMetric = StringMetricType(
            CommonMetricData(
                disabled = true,
                category = "telemetry",
                lifetime = Lifetime.APPLICATION,
                name = "stringMetric",
                sendInPings = listOf("store1")
            )
        )
        assertNull(stringMetric.testGetValue())
    }

    @Test
    fun `The API saves to secondary pings`() {
        // Define a 'stringMetric' string metric, which will be stored in "store1" and "store2"
        val stringMetric = StringMetricType(
            CommonMetricData(
                disabled = false,
                category = "telemetry",
                lifetime = Lifetime.APPLICATION,
                name = "string_metric",
                sendInPings = listOf("store1", "store2")
            )
        )

        // Record two strings of the same type, with a little delay.
        stringMetric.set("value")

        // Check that data was properly recorded in the second ping.
        assertEquals("value", stringMetric.testGetValue("store2"))

        stringMetric.set("overriddenValue")
        // Check that data was properly recorded in the second ping.
        assertEquals("overriddenValue", stringMetric.testGetValue("store2"))
    }

    @Test
    fun `Setting a long string records an error`() {
        // Define a 'stringMetric' string metric, which will be stored in "store1" and "store2"
        val stringMetric = StringMetricType(
            CommonMetricData(
                disabled = false,
                category = "telemetry",
                lifetime = Lifetime.APPLICATION,
                name = "string_metric",
                sendInPings = listOf("store1", "store2")
            )
        )

        stringMetric.set("0123456789".repeat(11))

        assertEquals(1, stringMetric.testGetNumRecordedErrors(ErrorType.INVALID_OVERFLOW))
    }
}
